core/ordering/cosipbft/mod.go

```diff
@@ -465,109 +476,128 @@ func (s *Service) doRound(ctx context.Context) error {
 
-			view, err := s.pbftsm.Expire(s.me)
-			if err != nil {
-				cancel()
-				return xerrors.Errorf("pbft expire failed: %v", err)
-			}
+	err = s.doPBFT(ctx)
+	if err != nil {
+		return xerrors.Errorf("pbft failed: %v", err)
+	}
 
-			viewMsg := types.NewViewMessage(view.GetID(), view.GetLeader(), view.GetSignature())
+	// The leader can be a new leader coming from a view change, so it resets
+	// the value as a round has finished.
+	s.failedRound = false
 
-			resps, err := s.rpc.Call(ctx, viewMsg, roster)
-			if err != nil {
-				cancel()
-				return xerrors.Errorf("rpc failed to send views: %v", err)
-			}
+	return nil
+}
 
-			for resp := range resps {
-				_, err = resp.GetMessageOrError()
-				if err != nil {
-					s.logger.Warn().Err(err).Str("to", resp.GetFrom().String()).Msg("view propagation failure")
-				}
-			}
+func (s *Service) doFollowerRound(ctx context.Context, roster authority.Authority) error {
+	// A follower has to wait for the new block, or the round timeout, to proceed.
+	select {
+	case <-time.After(s.timeoutRound):
+		if !s.roundHasFailed() {
+			return nil
+		}
 
-			statesCh := s.pbftsm.Watch(ctx)
+		s.logger.Info().Msg("round has failed, do a view change !")
 
-			state := s.pbftsm.GetState()
-			var more bool
+		s.pool.ResetStats() // avoid infinite view change
 
-			for state == pbft.ViewChangeState {
-				state, more = <-statesCh
-				if !more {
-					cancel()
-					return xerrors.New("viewchange failed")
-				}
+		view, err := s.pbftsm.Expire(s.me) // start the viewChange
+		if err != nil {
+			return xerrors.Errorf("pbft expire failed: %v", err)
+		}
+
+		viewMsg := types.NewViewMessage(view.GetID(), view.GetLeader(), view.GetSignature())
+
+		ctx, cancel := context.WithTimeout(ctx, s.timeoutRound)
+		defer cancel()
+
+		resps, err := s.rpc.Call(ctx, viewMsg, roster)
+		if err != nil {
+			return xerrors.Errorf("rpc failed to send views: %v", err)
+		}
+
+		for resp := range resps {
+			_, err = resp.GetMessageOrError()
+			if err != nil {
+				s.logger.Warn().Err(err).Str("to", resp.GetFrom().String()).Msg("view propagation failure")
 			}
+		}
 
-			s.logger.Debug().Msgf("view change successful for %d", viewMsg.GetLeader())
+		statesCh := s.pbftsm.Watch(ctx)
 
-			cancel()
-			return nil
-		case <-s.events:
-			// As a child, a block has been committed thus the previous view
-			// change succeeded.
-			s.failedRound = false
+		state := s.pbftsm.GetState()
+		var more bool
 
-			// A block has been created meaning that the round is over.
-			return nil
-		case <-s.closing:
-			return nil
+		for state == pbft.ViewChangeState {
+			state, more = <-statesCh
+			if !more {
+				return xerrors.New("view change failed")
+			}
 		}
-	}
 
-	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
+		s.logger.Info().Msgf("view change for %d", viewMsg.GetLeader())
 
-	s.logger.Debug().Uint64("index", s.blocks.Len()).Msg("round has started")
+		return nil
+	case <-s.events:
```

Comment:
It seems like readability / vertical spacing could be improved a little

```suggestion
		return nil
		
	case <-s.events:
```